kore/src/Kore/Step/Simplification/Overloading.hs

```diff
@@ -15,125 +19,236 @@ import Control.Error
-otherwise, return bottom, as the constructors are incompatible
+instance Diff UnifyOverloadingError
+
+instance Monoid UnifyOverloadingError where
+    mempty = NotApplicable
+
+type UnifyOverloading unifier a = ExceptT UnifyOverloadingError unifier (Pair a)
+
+notApplicable :: Monad unifier => UnifyOverloading unifier a
+notApplicable = empty
+
+throwBottom :: Monad unifier => String -> UnifyOverloading unifier a
+throwBottom = throwE . Clash
+
+{- |
+ If the two constructors form an overload pair, apply the overloading axioms
+ on the terms to make the constructors equal, then retry unification on them.
+
+See <https://github.com/kframework/kore/blob/master/docs/2019-08-27-Unification-modulo-overloaded-constructors.md>
 
  -}
 overloadedConstructorSortInjectionAndEquals
     :: (SimplifierVariable variable, MonadUnify unifier)
-    => HasCallStack
     => TermSimplifier variable unifier
     -> TermLike variable
     -> TermLike variable
     -> MaybeT unifier (Pattern variable)
-overloadedConstructorSortInjectionAndEquals
-    termMerger
-    first@(App_ firstHead _)
-    (Inj_ secondInj)
-  = do
-    OverloadSimplifier { isOverloaded } <- Simplifier.askOverloadSimplifier
-    Monad.guard (isOverloaded firstHead)
-    overloadedAndEqualsOverloading termMerger first secondInj
-overloadedConstructorSortInjectionAndEquals
-    termMerger
-    (Inj_ firstInj)
-    second@(App_ secondHead _)
+overloadedConstructorSortInjectionAndEquals termMerger firstTerm secondTerm
   = do
-    OverloadSimplifier { isOverloaded } <- Simplifier.askOverloadSimplifier
-    Monad.guard (isOverloaded secondHead)
-    overloadedAndEqualsOverloading termMerger second firstInj
-overloadedConstructorSortInjectionAndEquals
-    termMerger
-    first@(Inj_ inj@Inj { injTo, injChild = App_ firstHead firstChildren})
-    second@(Inj_ Inj { injChild = App_ secondHead secondChildren})
+    eunifier <- Monad.Trans.lift . runExceptT
+        $ unifyOverloading (Pair firstTerm secondTerm)
+    case eunifier of
+        Right (Pair firstTerm' secondTerm') -> Monad.Trans.lift $
+            termMerger firstTerm' secondTerm'
+        Left (Clash message) -> Monad.Trans.lift $
+            explainAndReturnBottom (fromString message) firstTerm secondTerm
+        Left NotApplicable -> empty
+
+{- |
+ Tests whether the pair of terms can be coerced to have the same constructors
+ at the top, and, if so, returns the thus obtained new pair.
+
+ See <https://github.com/kframework/kore/blob/master/docs/2019-08-27-Unification-modulo-overloaded-constructors.md>
+
+ If the overloading rules are not applicable, it would return @Nothing@.
+ However, if it detects an unification failure, it would return a reason
+ for the failure which can be used as an explanation for it.
+
+ Note: this algorithm is used for both matching and unification; that is why
+ the first and second terms in a pair are not interchangeable.
+-}
+unifyOverloading
```

Comment:
What happens if I feed the result of `unifyOverloading` back into itself? (That would probably be a good unit test.)

Reply:
I think it should probably return a `NotApplicable` answer, but we can definitely try it.